File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/HashData.java
package com.fincity.nocode.kirun.engine.function.system;

import com.fincity.nocode.kirun.engine.function.reactive.AbstractReactiveFunction;
import com.fincity.nocode.kirun.engine.model.*;
import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.fincity.nocode.kirun.engine.util.hash.HashUtil;
import com.google.gson.*;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Map;

public class HashData extends AbstractReactiveFunction {

    private static final String DEFAULT_ALGORITHM = "SHA-256";
    public static final String PARAMETER_DATA = "data";
    public static final String PARAMETER_ALGORITHM = "algorithm";
    public static final String PARAMETER_PRIMITIVE_LEVEL = "primitiveLevel";
    public static final String EVENT_RESULT_NAME = "result";

    private final FunctionSignature signature;

    public HashData() {
        this.signature = createSignature();
    }

    private static FunctionSignature createSignature() {
        return new FunctionSignature()
                .setNamespace(Namespaces.SYSTEM)
                .setName("HashData")
                .setParameters(Map.of(
                        PARAMETER_DATA, new Parameter()
                                .setSchema(Schema.ofAny(PARAMETER_DATA)),
                        PARAMETER_ALGORITHM, new Parameter()
                                .setSchema(Schema.ofString(PARAMETER_ALGORITHM).setEnums(List.of(
                                                new JsonPrimitive("SHA-256"),
                                                new JsonPrimitive("SHA-384"),
                                                new JsonPrimitive("SHA-512"),
                                                new JsonPrimitive("MD5"),
                                                new JsonPrimitive("MD2"),
                                                new JsonPrimitive("MD4"),
                                                new JsonPrimitive("SHA-1")))
                                        .setDefaultValue(new JsonPrimitive(DEFAULT_ALGORITHM))),
                        PARAMETER_PRIMITIVE_LEVEL, new Parameter()
                                .setSchema(Schema.ofBoolean(PARAMETER_PRIMITIVE_LEVEL)
                                        .setDefaultValue(new JsonPrimitive(false)))
                ))
                .setEvents(Map.ofEntries(Event.outputEventMapEntry(Map.of())));
    }

    @Override
    public FunctionSignature getSignature() {
        return this.signature;
    }

    @Override
    protected Mono<FunctionOutput> internalExecute(ReactiveFunctionExecutionParameters context) {

        Map<String, JsonElement> args = context.getArguments();

        String algorithm = args.get(PARAMETER_ALGORITHM).getAsString();

        JsonElement data = args.get(PARAMETER_DATA);

        boolean primitiveLevel = args.get(PARAMETER_PRIMITIVE_LEVEL).getAsBoolean();

        JsonElement hashValue = processElement(data, algorithm, primitiveLevel);

        return Mono.just(new FunctionOutput(List.of(EventResult.outputOf(Map.of(EVENT_RESULT_NAME, hashValue)))));
    }

    private JsonElement processElement(JsonElement element, String algorithm, boolean primitiveLevel) {

        if (element.isJsonNull()) {
            return new JsonPrimitive("null");
        }
        if (!primitiveLevel) {
            return HashUtil.createHash(element.toString(), algorithm);
        }

        if (element.isJsonPrimitive()) {
            JsonPrimitive primitive = element.getAsJsonPrimitive();
            if (primitive.isNumber()) {
                Number number = primitive.getAsNumber();
                return HashUtil.createHash(number.toString(), algorithm);
            } else if (primitive.isBoolean()) {
                return HashUtil.createHash(primitive.getAsBoolean(), algorithm);
            }
            return HashUtil.createHash(primitive.toString(), algorithm);
        }

        if (element.isJsonArray()) {
            JsonArray array = element.getAsJsonArray();
            JsonArray hashedArray = new JsonArray();
            array.forEach(e -> hashedArray.add(processElement(e, algorithm, true)));
            return hashedArray;
        }

        if (element.isJsonObject()) {
            JsonObject object = element.getAsJsonObject();
            JsonObject hashedObject = new JsonObject();
            object.entrySet()
                    .forEach(entry -> {
                        hashedObject.add(processElement(
                                        new JsonPrimitive(entry.getKey()), algorithm, true).getAsString(),
                                processElement(entry.getValue(), algorithm, true));
                    });
            return hashedObject;
        }

        return element;
    }

}

File: kirun-js/src/engine/function/system/HashData.ts
import { createHash } from 'crypto';
import { AbstractFunction } from '../AbstractFunction';
import { FunctionSignature } from '../../model/FunctionSignature';
import { Parameter } from '../../model/Parameter';
import { Schema } from '../../json/schema/Schema';
import { Event } from '../../model/Event';
import { Namespaces } from '../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../runtime/FunctionExecutionParameters';
import { FunctionOutput } from '../../model/FunctionOutput';
import { EventResult } from '../../model/EventResult';
import { MapUtil } from '../../util/MapUtil';

export class HashData extends AbstractFunction {
    private static readonly DEFAULT_ALGORITHM = 'sha256';
    public static readonly PARAMETER_DATA = 'data';
    public static readonly PARAMETER_ALGORITHM = 'algorithm';
    public static readonly PARAMETER_PRIMITIVE_LEVEL = 'primitiveLevel';
    public static readonly EVENT_RESULT_NAME = 'result';

    private readonly signature: FunctionSignature;

    public constructor() {
        super();
        this.signature = new FunctionSignature('Hash')
            .setNamespace(Namespaces.SYSTEM)
            .setParameters(
                new Map([
                    Parameter.ofEntry(
                        HashData.PARAMETER_DATA,
                        Schema.ofAny(HashData.PARAMETER_DATA)
                    ),
                    Parameter.ofEntry(
                        HashData.PARAMETER_ALGORITHM,
                        Schema.ofString(HashData.PARAMETER_ALGORITHM)
                            .setEnums([
                               'sha256',
                               'sha384',
                               'sha512',
                               'md5',
                               'md2',
                               'md4',
                               'sha1'])
                            .setDefaultValue(HashData.DEFAULT_ALGORITHM)
                    ),
                    Parameter.ofEntry(
                        HashData.PARAMETER_PRIMITIVE_LEVEL,
                        Schema.ofBoolean(HashData.PARAMETER_PRIMITIVE_LEVEL)
                            .setDefaultValue(false)
                    )
                ])
            )
            .setEvents(
                new Map([
                    [
                        Event.OUTPUT,
                        new Event(
                            Event.OUTPUT,
                            MapUtil.of(
                                HashData.EVENT_RESULT_NAME,
                                Schema.ofString(HashData.EVENT_RESULT_NAME)
                            )
                        )
                    ]
                ])
            );
    }

    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        const data = context.getArguments()?.get(HashData.PARAMETER_DATA);
        const algorithm = context.getArguments()?.get(HashData.PARAMETER_ALGORITHM) || HashData.DEFAULT_ALGORITHM;
        const primitiveLevel = context.getArguments()?.get(HashData.PARAMETER_PRIMITIVE_LEVEL);

        const result = primitiveLevel ? 
            this.processPrimitiveLevelData(data, algorithm) : 
            (data === null || data === undefined) ? 
                "null" : 
                createHash(algorithm).update(JSON.stringify(data)).digest('hex');

        return new FunctionOutput([
            EventResult.outputOf(new Map([[HashData.EVENT_RESULT_NAME, result]]))
        ]);
    }

    private processPrimitiveLevelData(value: any, algorithm: string): any {
        if (value === null || value === undefined) return "null";
        
        if (typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean') {
            return createHash(algorithm).update(JSON.stringify(value)).digest('hex');
        }

        if (Array.isArray(value)) {
            return value.map(item => this.processPrimitiveLevelData(item, algorithm));
        }

        if (typeof value === 'object') {
            const result: any = {};
            Object.entries(value).forEach(([key, val]) => {
                result[createHash(algorithm).update(JSON.stringify(key)).digest('hex')] = 
                    this.processPrimitiveLevelData(val, algorithm);
            });
            return result;
        }

        return createHash(algorithm).update(String(value)).digest('hex');
    }
}
